src/types/num.rs

```diff
@@ -243,43 +234,79 @@ mod slow_128bit_int_conversion {
 
             impl IntoPy<PyObject> for $rust_type {
                 fn into_py(self, py: Python) -> PyObject {
-                    let bytes = PyBytes::new(py, &self.to_le_bytes());
-                    let longtype = unsafe { PyType::from_type_ptr(py, &mut ffi::PyLong_Type) };
-                    longtype
-                        .call_method("from_bytes", (bytes, "little"), kwargs(py, $is_signed))
-                        .expect("Integer conversion (u128/i128 to PyLong) failed")
-                        .into_py(py)
+                    let lower = self as u64;
+                    let upper = (self >> SHIFT) as $half_type;
+                    unsafe {
+                        let shifted = PyObject::from_owned_ptr(
+                            py,
+                            ffi::PyNumber_Lshift(
+                                upper.into_py(py).as_ptr(),
+                                SHIFT.into_py(py).as_ptr(),
+                            ),
+                        );
+                        PyObject::from_owned_ptr(
+                            py,
+                            ffi::PyNumber_Or(shifted.as_ptr(), lower.into_py(py).as_ptr()),
+                        )
+                    }
                 }
             }
 
             impl<'source> FromPyObject<'source> for $rust_type {
                 fn extract(ob: &'source PyAny) -> PyResult<$rust_type> {
                     let py = ob.py();
-                    let num: &PyLong = unsafe {
-                        crate::FromPyPointer::from_owned_ptr_or_err(
+                    unsafe {
+                        let lower = err_if_invalid_value(
                             py,
-                            ffi::PyNumber_Index(ob.as_ptr()),
-                        )
-                    }?;
-                    let mut bytes = [0u8; std::mem::size_of::<$rust_type>()];
-                    let pybytes: &PyBytes = num
-                        .call_method("to_bytes", (bytes.len(), "little"), kwargs(py, $is_signed))?
-                        .extract()?;
-                    bytes.copy_from_slice(pybytes.as_bytes());
-                    Ok(<$rust_type>::from_le_bytes(bytes))
+                            -1 as _,
+                            ffi::PyLong_AsUnsignedLongLongMask(ob.as_ptr()),
```

Comment:
I think you'd probably want an assertion of some sort that `PyLong_UnsignedLongLongMask` returns the expected type, since C/C++ don't guarantee that it's `u64` -- it could be `u128` or some other type on an unusual OS.

Reply:
Hmm 🤔 , but even the Rust stdlib does not support it https://doc.rust-lang.org/nightly/std/os/raw/type.c_ulonglong.html.

Reply:
The docs are just showing the type used on the system they used to build the docs (`x86_64-unknown-linux-gnu` iirc), it doesn't mean it will always be `u64`.

Reply:
For an example, see https://doc.rust-lang.org/nightly/std/os/raw/type.c_long.html

Reply:
I meant the description: 
> The C standard technically only requires that this type be an unsigned integer with the size of a long long, although in practice, no system would have a long long that is not a u64, as most systems do not have a standardised u128 type.

Also, it has no `#[cfg` block:
![image](https://user-images.githubusercontent.com/16046705/102038723-549bef80-3e0b-11eb-8870-b7ffef1c7818.png)
So I think we don't need to consider the system where ull is u128 *for now*

Reply:
ok, yeah. I did check both the AS/400 and RISC-V 128-bit compilers (or at least their ABI proposals/docs) and they both have a 64-bit ull -- AS/400 just doesn't have a 128-bit int type (no intptr_t at all), and RV128 proposes `long long long` for 128-bit types.